src/openvpn/tun2tap.c

```diff
@@ -0,0 +1,196 @@
+                    if (arp_in->arp_command == htons(ARP_REPLY)){
+                        dmsg(D_TUN2TAP, "TUN2TAP: ignore arp reply");
+                        buf_clear(&c->c2.buf);
+                        ret = false;
+                    } else if (arp_in->arp_command == htons(ARP_REQUEST) ){
+                        if (arp_in->ip_dest == htonl(c->c1.tuntap->local)){
+                            // build reply
+                            // write arp to link
+                            struct openvpn_ethhdr hdr_out = {
+                                .proto=htons(OPENVPN_ETH_P_ARP)
+                            };
+                            struct openvpn_arp arp_out = { 0 };
+                            arp_out.mac_addr_type = htons(0x0001);
+                            arp_out.proto_addr_type = htons(0x0800);
+                            arp_out.mac_addr_size = 0x06;
+                            arp_out.proto_addr_size = 0x04;
+                            arp_out.arp_command = htons(ARP_REPLY);
+                            memcpy(arp_out.mac_src, c->options.lladdr_v, OPENVPN_ETH_ALEN);
+                            memcpy(arp_out.mac_dest, arp_in->mac_src, OPENVPN_ETH_ALEN);
+                            memcpy(hdr_out.source, c->options.lladdr_v, OPENVPN_ETH_ALEN);
+                            memcpy(hdr_out.dest, c->c1.tuntap->remote_mac_addr, OPENVPN_ETH_ALEN);
+                            arp_out.ip_src = arp_in->ip_dest;
+                            arp_out.ip_dest = arp_in->ip_src;
+                            buf_clear(&c->c2.buf);
+                            // dump_hex(BPTR(&c->c2.buf), BLEN(&c->c2.buf), "after clear");
+                            buf_write(&c->c2.buf, &hdr_out, sizeof(hdr_out));
+                            // dump_hex(BPTR(&c->c2.buf), BLEN(&c->c2.buf), "after write ether");
+                            buf_write(&c->c2.buf, &arp_out, sizeof(arp_out));
+                            // dump_hex(BPTR(&c->c2.buf), BLEN(&c->c2.buf), "after write arp");
+                            encrypt_sign(c, true);
+                            dmsg(D_TUN2TAP, "TUN2TAP: build arp reply success");
+                        } else {
+                            dmsg(D_TUN2TAP, "TUN2TAP: ignore any arp request not to me dest:%x me:%x", ntohl(arp_in->ip_dest), c->c1.tuntap->local);
+                            buf_clear(&c->c2.buf);
+                            ret = false;
+                        }
+                    } else{
+                        dmsg(D_TUN2TAP, "TUN2TAP: ignore uknown arp type: %x", ntohs(arp_in->arp_command));
+                        buf_clear(&c->c2.buf);
+                        ret = false;
+                    }
+                }
+            } else {
+                buf_advance(&c->c2.to_tun, sizeof(struct openvpn_ethhdr));
+            }
+        }
+    }
+    return ret;
+}
```

Comment:
Overall, I would like this code to be better documented with comments and also split into smaller function. E.g. `build_arp_reply` can/should be factored out.

Reply:
I think it's not beautiful to send too many arp params to that function, so i change the origin function name